### PR TITLE
CUDAWrappers::PreconditionIC and CUDAWrappers::PreconditionILU

### PR DESCRIPTION
**Pull request by masterleinad**
Builds on #7083.

This PR adds `IC` and `ILU` preconditioners for `CUDAWrappers::SparseMatrix` objects following the implementation in the (`cuSPARSE` documentation)[https://docs.nvidia.com/cuda/cusparse/index.html].

Currently, the preconditioners are only instantiated for `float` and `double` number types (in accordance with `CUDAWrappers::SparseMatrix` and `LinearAlgebra::CUDAWrappers::Vector`). I left the necessary template specializations in the source code nevertheless.

### CODE REVIEW

File: include/deal.II/lac/cuda_precondition.h [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+
+    /**
+     * Standardized data struct to pipe additional flags to the
+     * preconditioner.
+     */
+    struct AdditionalData
+    {
+      /**
+       * Constructor. cuSPARSE allows to compute and use level information.
+       * According to the documentation it is this might improve performance.
```
Comment (Rombur) [inline include/deal.II/lac/cuda_precondition.h]: Typo ` it is this might`

File: include/deal.II/lac/cuda_precondition.h [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+     *
+     * @note This function should only be called if the preconditioner has been
+     * initialized.
+     */
+    size_type
+    m() const;
+
+    /**
+     *  Return the dimension of the codomain (or range) space. Note that the
+     * matrix is square and has dimension $m \times m$.
```
Comment (Rombur) [inline include/deal.II/lac/cuda_precondition.h]: Wrong copy/paste

File: include/deal.II/lac/cuda_precondition.h [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+    const int *P_column_index_dev;
+
+    /**
+     * Pointer to the value (on the device) for a temporary (helper) vector
+     * used in vmult().
+     */
+    std::unique_ptr<Number[], void (*)(Number *)> tmp_dev;
+
+    /**
+     *
```
Comment (Rombur) [inline include/deal.II/lac/cuda_precondition.h]: No comment?

File: include/deal.II/lac/cuda_precondition.h [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+    /**
+     * The number of non-zero elements is the same as for the matrix this
+     * object has been initialized with.
+     */
+    int n_nonzero_elements;
+  };
+
+  /**
+   * This class implements an incomplete LU factorization preconditioner for
+   * @em symmetric CUDAWrappers::SparseMatrix matrices.
```
Comment (Rombur) [inline include/deal.II/lac/cuda_precondition.h]: I don't think that the matrix needs to be symmetric

File: include/deal.II/lac/cuda_precondition.h [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+
+    /**
+     * Apply the preconditioner.
+     */
+    void
+    vmult(LinearAlgebra::CUDAWrappers::Vector<Number> &      dst,
+          const LinearAlgebra::CUDAWrappers::Vector<Number> &src) const;
+
+    /**
+     * Apply the preconditioner. Since the preconditioner is symmetric, this
```
Comment (Rombur) [inline include/deal.II/lac/cuda_precondition.h]: copy/paste ?

File: include/deal.II/lac/cuda_precondition.h [hunk 6]
```diff
@@ -0,0 +1,523 @@
...
+     * matrix is square and has dimension $m \times m$.
+     *
+     * @note This function should only be called if the preconditioner has been
+     * initialized.
+     */
+    size_type
+    m() const;
+
+    /**
+     *  Return the dimension of the codomain (or range) space. Note that the
```
Comment (Rombur) [inline include/deal.II/lac/cuda_precondition.h]: Wrong copy paste

File: include/deal.II/lac/cuda_precondition.h [hunk 7]
```diff
@@ -0,0 +1,523 @@
...
+    const int *P_column_index_dev;
+
+    /**
+     * Pointer to the value (on the device) for a temporary (helper) vector
+     * used in vmult().
+     */
+    std::unique_ptr<Number[], void (*)(Number *)> tmp_dev;
+
+    /**
+     *
```
Comment (Rombur) [inline include/deal.II/lac/cuda_precondition.h]: No comment?

File: source/lac/cuda_precondition.cu [hunk 8]
```diff
@@ -0,0 +1,1844 @@
...
+
+    n_rows             = A.m();
+    n_nonzero_elements = A.n_nonzero_elements();
+    AssertDimension(A.m(), A.n());
+
+    matrix_pointer                      = &A;
+    const auto          cusparse_matrix = A.get_cusparse_matrix();
+    const Number *const A_val_dev       = std::get<0>(cusparse_matrix);
+
+    // create a copy of the matrix entries
```
Comment (Rombur) [inline source/lac/cuda_precondition.cu]: Why do you do this? Is the data overwritten?

File: source/lac/cuda_precondition.cu [hunk 9]
```diff
@@ -0,0 +1,1844 @@
...
+    P_val_dev.reset(allocate_device_vector<Number>(n_nonzero_elements));
+    cudaError_t cuda_status            = cudaMemcpy(P_val_dev.get(),
+                                         A_val_dev,
+                                         n_nonzero_elements * sizeof(Number),
+                                         cudaMemcpyDeviceToDevice);
+    P_column_index_dev                 = std::get<1>(cusparse_matrix);
+    P_row_ptr_dev                      = std::get<2>(cusparse_matrix);
+    const cusparseMatDescr_t mat_descr = std::get<3>(cusparse_matrix);
+
+    // initializa an internal buffer we need later on
```
Comment (Rombur) [inline source/lac/cuda_precondition.cu]: Typo

File: source/lac/cuda_precondition.cu [hunk 10]
```diff
@@ -0,0 +1,1844 @@
...
+
+    const Number *const src_dev = src.get_values();
+    Number *const       dst_dev = dst.get_values();
+    // step 6: solve L*z = alpha*x
+    const Number     alpha = internal::NumberType<Number>::value(1.);
+    cusparseStatus_t status =
+      cusparseXcsrsv2_solve(cusparse_handle,
+                            CUSPARSE_OPERATION_NON_TRANSPOSE,
+                            n_rows,
+                            n_nonzero_elem
```
Comment (Rombur) [inline source/lac/cuda_precondition.cu]: Typo

Comment (Rombur) [inline source/lac/cuda_precondition.cu]: This is not what the documentation says.

File: source/lac/cuda_precondition.cu (hunk 8 above)
Comment (masterleinad) [reply source/lac/cuda_precondition.cu]: Yes, it is.

File: source/lac/cuda_precondition.cu (hunk 10 above)
Comment (masterleinad) [reply source/lac/cuda_precondition.cu]: What do you mean?

Comment (masterleinad) [reply source/lac/cuda_precondition.cu]: What do you mean?

Comment (Rombur) [reply source/lac/cuda_precondition.cu]: It looks like there is a problem with github :( This should be at line 1828 in the `Tvmult` function. The documentation says that the function exists.

Comment (Rombur) [reply source/lac/cuda_precondition.cu]: This should be at line 1648 I don't why it shows the wrong line

**Comment by masterleinad** [comment]
@rombur I think I addressed all your comments.

/run-tests

**Pull request merged**